File: backends/go/runtime/invoke.go
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael

package sce

import (
	"fmt"
	"strings"
)

// PendingInvoke represents a W3C SCXML 6.4 pending invoke structure for the
// defer/cancel/execute pattern.
//
// Ports Rust PendingInvoke from backends/rust/runtime/src/invoke.rs.
type PendingInvoke[S comparable] struct {
	// InvokeID is the W3C SCXML 6.4.1 runtime-generated invoke identifier.
	// Format: "stateid.platformid.index" (e.g., "s01.140234567890._invoke_0")
	InvokeID string

	// State is the state containing the invoke element.
	State S
}

// ChildSession represents a W3C SCXML 6.4 active child session.
//
// Ports Rust ChildSession from backends/rust/runtime/src/invoke.rs.
type ChildSession struct {
	// SessionID is the child's session ID (format: "parentSessionId.invokeId").
	SessionID string

	// InvokeID is the invoke element ID (e.g., "_invoke_0").
	InvokeID string

	// ParentSessionID is the parent's session ID.
	ParentSessionID string

	// Autoforward indicates whether to autoforward events to child (W3C SCXML 6.4.6).
	Autoforward bool

	// FinalizeScript is the <finalize> handler script content (W3C SCXML 6.5).
	FinalizeScript string
}

// DeferInvoke adds a pending invoke for deferred execution at macrostep end
// (W3C SCXML 6.4).
//
// Ports Rust defer_invoke from backends/rust/runtime/src/invoke.rs.
func DeferInvoke[S comparable](pending *[]PendingInvoke[S], invoke PendingInvoke[S]) {
	*pending = append(*pending, invoke)
}

// CancelInvokesForState removes pending invokes for an exited state
// (W3C SCXML 6.4).
//
// Ports Rust cancel_invokes_for_state from backends/rust/runtime/src/invoke.rs.
func CancelInvokesForState[S comparable](pending *[]PendingInvoke[S], state S) {
	n := 0
	for _, p := range *pending {
		if p.State != state {
			(*pending)[n] = p
			n++
		}
	}
	*pending = (*pending)[:n]
}

// ExecutePendingInvokes executes all pending invokes at macrostep end
// (W3C SCXML 6.4). Uses copy-and-clear pattern to prevent iterator
// invalidation during execution.
//
// Ports Rust execute_pending_invokes from backends/rust/runtime/src/invoke.rs.
func ExecutePendingInvokes[S comparable](pending *[]PendingInvoke[S], executor func(*PendingInvoke[S])) {
	if len(*pending) == 0 {
		return
	}

	// W3C SCXML 6.4: Copy pending list to prevent iterator invalidation
	invokesToExecute := make([]PendingInvoke[S], len(*pending))
	copy(invokesToExecute, *pending)
	*pending = (*pending)[:0]

	for i := range invokesToExecute {
		executor(&invokesToExecute[i])
	}
}

// CreateDoneInvokeEventName creates a done.invoke event name (W3C SCXML 6.3.1).
//
// Ports Rust create_done_invoke_event_name from backends/rust/runtime/src/invoke.rs.
func CreateDoneInvokeEventName(invokeID string) string {
	return fmt.Sprintf("done.invoke.%s", invokeID)
}

// IsValidInvokeID validates invoke ID format (W3C SCXML 3.12.1).
//
// Ports Rust is_valid_invoke_id from backends/rust/runtime/src/invoke.rs.
func IsValidInvokeID(invokeID string) bool {
	return invokeID != ""
}

// GetPendingCount returns the count of pending invokes (W3C SCXML 6.4).
func GetPendingCount[S comparable](pending []PendingInvoke[S]) int {
	return len(pending)
}

// IsInvokePending checks if a specific invoke is pending (W3C SCXML 6.4).
func IsInvokePending[S comparable](pending []PendingInvoke[S], invokeID string) bool {
	for _, p := range pending {
		if p.InvokeID == invokeID {
			return true
		}
	}
	return false
}

// ChildEngine is the interface that parent state machines use to interact with
// child invoke sessions (W3C SCXML 6.4). Generated code implements this for
// each concrete child state machine type.
type ChildEngine interface {
	Initialize()
	Tick()
	IsInFinalState() bool
	RaiseExternalByName(eventName, eventData string)
	SetCompletionCallback(callback func())
	GetParentEventQueue() chan ParentEvent

	// DonedataAtFinal returns the donedata payload stashed by the child's
	// top-level <final> onentry, for the parent's RaiseDoneInvoke to lift
	// onto done.invoke.<id>._event.data (W3C SCXML 5.5 + 6.3.1).
	DonedataAtFinal() string
}

// DrainAndRaiseChildEvents drains events from a child's ParentExternalQueue
// and raises them in the parent engine with proper metadata (W3C SCXML 6.4).
// Port of Rust drain_and_raise_child_events().
func DrainAndRaiseChildEvents[S comparable, E comparable](
	childEngine ChildEngine,
	activeInvokes map[string]*ChildSession,
	invokeKey string,
	engine *Engine[S, E],
) {
	if childEngine == nil {
		return
	}
	queue := childEngine.GetParentEventQueue()
	if queue == nil {
		return
	}
	for {
		select {
		case ev := <-queue:
			event, ok := engine.policy.GetEventFromName(ev.Name)
			if !ok {
				continue
			}
			meta := NewEventWithMetadata(event)
			meta.Metadata.Data = ev.Data
			if cs, found := activeInvokes[invokeKey]; found {
				meta.Metadata.InvokeID = cs.InvokeID
				meta.Metadata.Origin = cs.SessionID
				meta.Metadata.OriginType = SCXMLEventProcessorType
			}
			engine.RaiseExternalWithMeta(meta)
		default:
			return
		}
	}
}

// RaiseDoneInvoke raises a done.invoke event in the parent engine (W3C SCXML 6.3.1).
// Port of Rust raise_done_invoke().
//
// donedata is the child's stashed top-level <final> donedata payload
// (W3C SCXML 5.5 + 6.3.1); it is threaded onto EventMetadata.Data so
// the parent's transition cond can read _event.data. Pass "" when the
// child's <final> carries no donedata.
func RaiseDoneInvoke[S comparable, E comparable](
	invokeID string,
	donedata string,
	engine *Engine[S, E],
) {
	eventName := "done.invoke." + invokeID
	event, ok := engine.policy.GetEventFromName(eventName)
	if !ok {
		// Try without the specific ID (generic done.invoke)
		event, ok = engine.policy.GetEventFromName("done.invoke")
		if !ok {
			return
		}
	}
	meta := NewEventWithMetadata(event)
	meta.Metadata.InvokeID = invokeID
	meta.Metadata.Data = donedata
	engine.RaiseExternalWithMeta(meta)
}

// IsPlatformEvent checks if an event name is a platform event that should NOT
// be autoforwarded (W3C SCXML 6.4.6).
func IsPlatformEvent(name string) bool {
	return strings.HasPrefix(name, "#_")
}

File: backends/python/runtime/sce_runtime/invoke.py
# SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
# SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
"""W3C SCXML 6.4 invoke runtime — child-session lifecycle, autoforward,
finalize, and done.invoke wiring.

`Invoke[E]` is the ABC every concrete invoke kind implements. `ScxmlInvoke`
covers the in-process `<invoke type="scxml">` case (child is a sibling
generated `*_sm.py` module) — the only kind γ-4a lights up. HTTP transports
(`BasicHTTPEventProcessor`) ride a different `Invoke` subclass that γ-4b
lands; mesh-rpc invokes stay permanently rejected per the C++-first mesh
policy.

The runtime keeps the state on `Engine` (`_pending_invokes`,
`_active_invokes`); generated policy code populates those lists via the
`defer_invokes_on_entry` / `execute_pending_invokes` hooks. This split
mirrors Go's `sce.PendingInvoke` / `ChildSession` + policy method
contract (`backends/go/runtime/invoke.go`) so the two backends evolve
together when W3C clarifies invoke semantics.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Generic, Iterable, List, Optional, Tuple, TypeVar

E = TypeVar("E")


@dataclass
class PendingInvoke:
    """W3C SCXML 6.4 — a `<invoke>` whose owner state has been entered but
    whose child has not yet been instantiated. The engine drains the
    pending list AFTER the current macrostep settles (defer-execute
    pattern), so onentry handlers always observe a stable configuration
    before any child starts running."""

    invoke_id: str
    owner_state_name: str


@dataclass
class ChildSession:
    """Bookkeeping for an active `<invoke>` session (W3C SCXML 6.4)."""

    invoke_id: str
    autoforward: bool = False
    finalize_data: Any = None


class Invoke(ABC, Generic[E]):
    """W3C SCXML 6.4 abstract invoke.

    Subclasses bridge the parent engine to whatever the invoke target
    actually is — another statechart (`ScxmlInvoke`), an HTTP endpoint
    (γ-4b), or any external transport. Every concrete `Invoke`:
      - starts the target on `start(parent_engine)`,
      - drains target-to-parent events through `drain_events()`,
      - forwards parent-to-target events through `forward_event(...)`,
      - reports completion via `is_done()` and `done_data()`.

    The engine never reaches into the target directly — every interaction
    flows through these five operations, which keeps the parent's
    `advance_time` / macrostep loop agnostic of the invoke kind.
    """

    @abstractmethod
    def start(self, parent_engine) -> None:
        """W3C SCXML 6.4 — instantiate and initialise the target."""

    @abstractmethod
    def tick(self) -> None:
        """W3C SCXML 6.4 — advance the target by one macrostep
        equivalent. Idempotent when the target is already done."""

    @abstractmethod
    def is_done(self) -> bool:
        """True once the target has reached its terminal configuration
        (W3C 3.7 top-level `<final>` for SCXML children)."""

    @abstractmethod
    def drain_events(self) -> Iterable[Tuple[str, Any]]:
        """W3C SCXML 6.4 — yield every `(event_name, data)` the target
        has raised to its parent (e.g. `<send target="#_parent">`)
        since the last drain. The runtime promotes each tuple onto
        the parent's external queue via `Engine.send_external_by_name`."""

    @abstractmethod
    def forward_event(self, event_name: str, data: Any) -> None:
        """W3C SCXML 6.4.1 — deliver an autoforwarded event into the
        target. The descriptor matches the wire name (`done.foo`)
        rather than the parent's `Event` enum, so the implementation
        is responsible for any local name→Event resolution."""

    @abstractmethod
    def cancel(self) -> None:
        """W3C SCXML 6.4 — terminate the target (parent state exit /
        engine shutdown). After this the engine drops the entry from
        `_active_invokes`; subsequent ticks are skipped."""

    def done_data(self) -> Any:
        """W3C SCXML 5.5 + 6.3.1 — payload captured at the target's
        terminal `<final>`. Surfaces on `done.invoke.<id>._event.data`
        in the parent. Default `None` — subclasses with no donedata
        path keep the default."""
        return None


class ScxmlInvoke(Invoke):
    """W3C SCXML 6.4 — in-process child statechart invoke. Wraps a
    pre-instantiated child `Engine` plus a parent-event queue the
    child's generated policy writes into when it executes
    `<send target="#_parent">`.

    The child's main loop is its own `advance_time(0)` macrostep settler
    — `tick()` runs that so any newly-due child schedules drain into
    the child's queues before the parent observes the child's parent-
    bound events. Forwarding and cancellation route through the
    child's `Engine` API directly; no extra metadata is invented."""

    def __init__(self, child_engine, parent_event_queue: List[Tuple[str, Any]]):
        self._child = child_engine
        self._parent_queue = parent_event_queue
        self._started = False

    def start(self, parent_engine) -> None:
        if not self._started:
            self._child.initialize()
            self._started = True

    def tick(self) -> None:
        # Drain due child schedules + run the child's macrostep so any
        # parent-bound events emitted during this iteration land in
        # `_parent_queue` before the parent's drain_events() call.
        if self._child.is_running and not self._child.reached_final:
            self._child.advance_time(0)

    def is_done(self) -> bool:
        return self._child.reached_final

    def drain_events(self) -> Iterable[Tuple[str, Any]]:
        while self._parent_queue:
            yield self._parent_queue.pop(0)

    def forward_event(self, event_name: str, data: Any) -> None:
        # The child policy's name→Event resolver lifts the string onto
        # the child's Event enum. Unknown names silently drop, matching
        # W3C 5.10.1 ("if no transition is enabled the event is lost").
        event = self._child.policy.get_event_from_name(event_name)
        if event is None:
            return
        # W3C SCXML C.1: parent→child delivery rides the SCXML Event
        # I/O Processor, so the child sees `_event.origintype` = SCXML
        # processor URI (test253). Imported lazily to avoid a cycle.
        from .engine import SCXML_EVENT_PROCESSOR_URI
        from .event import EventMetadata
        self._child.send_event(
            event,
            EventMetadata(
                event_type="external",
                data=data,
                origin_type=SCXML_EVENT_PROCESSOR_URI,
            ),
        )

    def cancel(self) -> None:
        # W3C SCXML 6.4.2 — terminate the child. Marking the engine
        # stopped is enough to keep its scheduler from delivering any
        # remaining `<send delay>` entries (advance_time gates on
        # `is_running and not reached_final`).
        self._child.stop()

    def done_data(self) -> Any:
        return getattr(self._child, "done_data", None)

    @property
    def child(self):
        """Access to the underlying child engine — only for the policy
        hooks that need to introspect child config (e.g. autoforward
        resolution). The engine itself reaches the child only through
        the five Invoke methods above."""
        return self._child


def is_platform_event(event_name: str) -> bool:
    """W3C SCXML 6.4.1 — platform events (prefix `#_`) must not be
    autoforwarded. Mirrors `sce.IsPlatformEvent` in Go."""
    return event_name.startswith("#_")


def create_done_invoke_event_name(invoke_id: str) -> str:
    """W3C SCXML 6.3.1 — `done.invoke.<id>` event name."""
    return f"done.invoke.{invoke_id}"
